packages/jest-circus/src/run.ts

```diff
@@ -28,27 +28,52 @@ import {
 // the original values in the variables before we require any files.
 const {setTimeout} = globalThis;
 
-type ConcurrentTestEntry = Omit<Circus.TestEntry, 'fn'> & {
-  fn: Circus.ConcurrentTestFn;
-  done: Promise<void>;
-};
+const testNameStorage = new AsyncLocalStorage<string>();
 
 const run = async (): Promise<Circus.RunResult> => {
   const {rootDescribeBlock, seed, randomize} = getState();
+  jestExpect.setState({
+    currentConcurrentTestName: () => testNameStorage.getStore(),
+  });
   const rng = randomize ? rngBuilder(seed) : undefined;
   await dispatch({name: 'run_start'});
-  await _runTestsForDescribeBlock(rootDescribeBlock, rng, true);
+  await _runTestsForDescribeBlock(rootDescribeBlock, rng);
   await dispatch({name: 'run_finish'});
   return makeRunResult(
     getState().rootDescribeBlock,
     getState().unhandledErrors,
   );
 };
 
+function* regroupConcurrentChildren(
+  children: Array<Circus.DescribeBlock | Circus.TestEntry>,
+) {
+  const concurrentTests = children.filter(
+    (child): child is Circus.TestEntry =>
+      child.type === 'test' && child.concurrent,
+  );
+  if (concurrentTests.length === 0) {
+    yield* children;
+    return;
+  }
+  let collectedConcurrent = false;
+  for (const child of children) {
+    if (child.type === 'describeBlock') {
+      yield child;
+    } else if (child.concurrent) {
+      if (!collectedConcurrent) {
+        collectedConcurrent = true;
+        yield {tests: concurrentTests, type: 'test-concurrent' as const};
+      }
+    } else {
+      yield child;
+    }
```

Comment:
Would be more readable, this extra `if` block doesn't seem to be needed
```suggestion
    if (child.concurrent) {
      if (!collectedConcurrent) {
        collectedConcurrent = true;
        yield {tests: concurrentTests, type: 'test-concurrent' as const};
      }
    } else {
      yield child;
    }
```